pyts/classification/bosssp.py

```diff
@@ -0,0 +1,270 @@
+        X : array-like, shape = (n_samples, n_timestamps)
+            Training vector.
+
+        y : array-like, shape = (n_samples,)
+            Class labels for each data sample.
+
+        Returns
+        -------
+        self : object
+
+        """
+        length = len(X)
+        words = self._boss_word_extractor(X, y)
+        if self.level >= 2:
+            words_2 = self._boss_word_extractor(X[:length//2],
+                                                y[:length//2])
+            words_3 = self._boss_word_extractor(X[length//2:],
+                                                y[:length//2:])
+            words = words + words_2 + words_3
+        if self.level == 3:
+            words_4 = self._boss_word_extractor(X[:length//4],
+                                                y[:length//4])
+            words_5 = self._boss_word_extractor(X[length//4:length//2],
+                                                y[length//4:length//2])
+            words_6 = self._boss_word_extractor(X[length//2:length*3//4],
+                                                y[length//2:length*3//4])
+            words_7 = self._boss_word_extractor(X[length*3//4:],
+                                                y[length*3//4:])
+            words = words + words_4 + words_5 + words_6 + words_7
+        word_count = dict(Counter(words))
+        sorted_word_count = {key: value for key, value in sorted(
+                             word_count.items())}
+        self._word_count = sorted_word_count
+        return self
+
+    def _boss_word_extractor(self, X, y):
+        X, y = check_X_y(X, y)
+        n_samples, n_timestamps = X.shape
+        check_classification_targets(y)
+
+        window_size, window_step = self._check_params(n_timestamps)
+        n_windows = (n_timestamps - window_size + window_step) // window_step
+
+        X_windowed = _windowed_view(
+            X, n_samples, n_timestamps, window_size, window_step
+        )
+        X_windowed = X_windowed.reshape(n_samples * n_windows, window_size)
+
+        sfa = SymbolicFourierApproximation(
+            n_coefs=self.word_size, drop_sum=self.drop_sum, anova=self.anova,
+            norm_mean=self.norm_mean, norm_std=self.norm_std,
+            n_bins=self.n_bins, strategy=self.strategy, alphabet=self.alphabet
+        )
+        y_repeated = np.repeat(y, n_windows)
+        X_sfa = sfa.fit_transform(X_windowed, y_repeated)
+
+        X_word = np.asarray([''.join(X_sfa[i])
+                             for i in range(n_samples * n_windows)])
+        X_word = X_word.reshape(n_samples, n_windows)
+
+        if self.numerosity_reduction:
+            not_equal = np.c_[X_word[:, 1:] != X_word[:, :-1],
+                              np.full(n_samples, True)]
+            X_bow = np.asarray([' '.join(X_word[i, not_equal[i]])
+                                for i in range(n_samples)])
+        else:
+            X_bow = np.asarray([' '.join(X_word[i]) for i in range(n_samples)])
```

Comment:
The code is very similar to the `fit` method of BOSS, which is normal, but I wonder if it would not be better to use it instead of having extremely similar code. This would also mean that fewer tests are required.